### ISSUE TITLE
2.0.0a4 L003 line break within datatype

### ISSUE THREAD

**Issue by fmms** [sqlfluff/sqlfluff#4328]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I tried to format the following tsql statement:
```sql
DECLARE @SqlTruncate nvarchar(400) = IIF(@myvar = 123 OR @myvar = 2
, 'yes'
, 'no'
)

```

### Expected Behaviour

I expected the statement to stay as is.

### Observed Behaviour

In fact the SQL was formatted the following way:
```sql
DECLARE @SqlTruncate nvarchar(
  400) = IIF(@myvar = 123 OR @myvar = 2
, 'yes'
, 'no'
)
```

### How to reproduce

```
PS > cat .\q1.sql
DECLARE @SqlTruncate nvarchar(400) = IIF(@myvar = 123 OR @myvar = 2
, 'yes'
, 'no'
)
PS > sqlfluff.exe fix --dialect tsql .\q1.sql
==== finding fixable violations ====
== [q1.sql] FAIL
L:   1 | P:  31 | L003 | Expected line break and indent of 2 spaces before '400'.
L:   2 | P:   1 | L003 | Expected indent of 2 spaces.
L:   3 | P:   1 | L003 | Expected indent of 2 spaces.
==== fixing violations ====
3 fixable linting violations found
Are you sure you wish to attempt to fix these? [Y/n] ...
Attempting fixes...
Persisting Changes...
== [q1.sql] PASS
Done. Please check your files to confirm.
All Finished 📜 🎉!
PS C:\Users\fmoeller\Desktop\RepositoryVorlage>
```

### Dialect

tsql

### Version

```
==== sqlfluff ====
sqlfluff:              2.0.0a4 python:                 3.11.0
implementation:        cpython verbosity:                   2
dialect:                  tsql templater:                 raw
rules:                                all
```

### Configuration

default

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by barrywhart** [comment]
This slightly altered file makes a better test case, because it only generates a single fix, the problematic one:
```
DECLARE @SqlTruncate nvarchar(400) = IIF(@myvar = 123 OR @myvar = 2,
    'yes',
    'no'
)
```

**Comment by barrywhart** [comment]
This SQL file produces the same behavior:
```
DECLARE @SqlTruncate nvarchar(400) = FN(TRUE
)
```

If I change it to this, there's no lint message and `fix` makes no changes:
```
DECLARE @SqlTruncate nvarchar(400) = FN(
    TRUE
)
```

Therefore, I think the bug is that it _should_ add a newline and indent after `IIF(`, but it's adding it after `nvarchar(` instead.

**Comment by barrywhart** [comment]
The undesired change is being created here:
```
> /Users/barry.hart/dev/sqlfluff/src/sqlfluff/utils/reflow/elements.py(543)indent_to()
                idx = self.segments.index(ws_seg)
                if not description:
                    # Prefer before, because it makes the anchoring better.
                    if before:
                        import pdb; pdb.set_trace()
                        description = (
                            "Expected line break and "
                            f"{_indent_description(desired_indent)} "
                            f"before {before.raw!r}."
                        )
```

But I think the actual bug lies pretty far upstream, based on this:
```
[_IndentPoint(idx=1, indent_impulse=1, indent_trough=0, initial_indent_balance=0, last_line_break_idx=None, is_line_break=False, untaken_indents=()),
 _IndentPoint(idx=7, indent_impulse=1, indent_trough=0, initial_indent_balance=1, last_line_break_idx=None, is_line_break=False, untaken_indents=(1,)),
 _IndentPoint(idx=9, indent_impulse=-1, indent_trough=-1, initial_indent_balance=2, last_line_break_idx=None, is_line_break=False, untaken_indents=(1, 2)),
 _IndentPoint(idx=17, indent_impulse=1, indent_trough=0, initial_indent_balance=1, last_line_break_idx=None, is_line_break=False, untaken_indents=(1,)),
 _IndentPoint(idx=33, indent_impulse=0, indent_trough=0, initial_indent_balance=2, last_line_break_idx=None, is_line_break=True, untaken_indents=(1, 2))]
```

Notice the `indent_impulse` values: 1, 1, -1, 1, 0. I think the first `_IndentPoint` should not be there.

**Comment by barrywhart** [comment]
`ReflowPoint.indent_to()` is being called with `before=<CodeSegment: ([L:  1, P:  9]) '@SqlTruncate'>`.

**Comment by barrywhart** [comment]
This portion of the docstring for `lint_indent_points()` describes the intended behavior:
```
    1. Any untaken indents. An untaken indent is only valid if its
    corresponding dedent is on the same line. If that is not the case,
    there should be a line break at the location of the indent and dedent.
```

As noted in the previous issue comment, I don't think the code is actually doing this. Instead, if an element has `newline, dedent`, it's acting like the dedent occurred on that line, but it really occurred on the **next** line.

I'm going to try and account for this by adding new field(s) to `_IndentPoint`. The existing fields won't change meaning, but the new fields will reflect the indent impulse, etc. **only prior to** any newline(s) in the element. 

Once those fields are added, we can modify `reindent._lint_line_untaken_positive_indents()` to use this additional information to handle the `newline, dedent` situation correctly.

**Comment by alanmcruickshank** [comment]
So I think this is another example of the same issue described in #4234. Perhaps made worse by an additional matching bug in the process.

There's already a suggested fix in the linked issue, and I suggest we implement that fix first, and then revisit this one to find out if the additional matching complexity here remains at that stage.

@barrywhart if you want to take a go at #4234, then you're very welcome - but it could be a messy one. Otherwise that one is next on my radar after getting a few more steps with the rules reorg.

**Comment by alanmcruickshank** [comment]
@barrywhart - on the matching problem you describe - I think the solution in this case is that the iteration in `_lint_line_untaken_positive_indents`, i.e. the part starting...
```python
# The closing indent balance *does* correspond to an
# untaken indent on this line. We *should* force a newline
# at that position.
for ip in indent_points:
   ...
```

Should instead search in reverse (i.e. `for ip in reversed(indent_points):`), so that it finds the last point on the line matching the criteria rather than the last. I experimented with this as a fix for one of the other recent bugs, but I seem to remember the change having some unexpected knock on effects and eventually abandoned it as not the right fix for the issue I was trying to solve at the time.

I'm still convinced that a reverse search in that place would help with issues like this.

**Comment by alanmcruickshank** [comment]
With the fox proposed in #4420, the new behaviour is to fix the query in the original post to:

```sql
-- original:
DECLARE @SqlTruncate nvarchar(400) = IIF(@myvar = 123 OR @myvar = 2
, 'yes'
, 'no'
);
-- fixed:
DECLARE @SqlTruncate nvarchar(400) = IIF(
    @myvar = 123 OR @myvar = 2
    , 'yes'
    , 'no'
);
```

It also fixes the modified query posted by @barrywhart to:

```sql
-- original:
DECLARE @SqlTruncate nvarchar(400) = IIF(@myvar = 123 OR @myvar = 2,
    'yes',
    'no'
)
-- fixed:
DECLARE @SqlTruncate nvarchar(400) = IIF(
    @myvar = 123 OR @myvar = 2,
    'yes',
    'no'
)
```

Helpfully it looks like at least the new fix is at least consistent. 👍 

@barrywhart @fmms - can we also say that the open PR resolves this issue or is there more to do to also resolve this one?

**Comment by fmms** [comment]
Thanks @alanmcruickshank looks very good to me

**Comment by barrywhart** [comment]
Looks good to me too. I'll update the PR description to close this issue as well.
